### ISSUE TITLE
Add a hint: Unnecessary import of dart:async

### ISSUE THREAD

**Issue by matanlurey** [dart-lang/sdk#34401]
As of https://github.com/dart-lang/sdk/commit/f0172a4c417ef9515a9faf54084119548daa52c2, `Future` and `Stream` are now exported via `dart:core` - i.e. are always available in the default namespace unless you import `dart:core` with `show/hide` or a prefix (`as`).

Lots of code is going to have this pattern for historic reasons now:

```dart
import 'dart:async';

Future<String> helloAsync() async => 'Hello';
```

... I don't believe this will be considered an `UNUSED_IMPORT`, because it is _used_, it just isn't _necessary_ (the symbol is always exported via `dart:core`, now). It would be nice to make this either a lint or a hint with an auto-fix to remove it:

```dart
// HINT: Unnecessary import. Future and Stream are available via dart:core.
import 'dart:async';

Future<String> helloAsync() async => 'Hello';
```

I don't think it is safe to _generalize_ this, at least not without a broader discussion, because I imagine there are packages out there that export the same symbols/classes/functions from multiple entry-points intentionally.

**Comment by natebosch** [comment]
> I don't think it is safe to generalize this, at least not without a broader discussion, because I imagine there are packages out there that export the same symbols/classes/functions from multiple entry-points intentionally.

I don't see why it wouldn't be safe as long as its a hint in my editor without failing the build.

**Comment by matanlurey** [comment]
@natebosch:

> I don't see why it wouldn't be safe as long as its a hint in my editor without failing the build.

I suppose so. I think the main issue is how do we surface this requirement for continuous integration. For example, do we want to add some sort of `package:meta` annotation, `@PreferImportFrom`:

```dart
// lib/a.dart
import 'package:meta/meta.dart';

@PreferImportFrom('a.dart')
class A {}
```

```dart
// lib/b.dart

export 'a.dart' show A;
```

```dart
// lib/c.dart

// HINT: A should be imported from a.dart. <QUICK FIX>
import 'b.dart' show A;
```

-----

... otherwise we can't make this a requirement internally, nor on Travis, without defining how severe the infraction is (should block CLs/PRs, or just be a nuisance like `@Deprecated()`). I also imagine Flutter will care about that distinction.

(There is a longer meta discussion around whether we should have "CI-mode" lints/hints that are only triggered on _new_ code, similar to our internal infra, but that's probably out of scope here - maybe we should open another issue)

**Comment by kevmoo** [comment]
If this lint is ever created, it **MUST** be aware of the SDK constraint on the associated package – and require the right lower bound.

**Comment by matanlurey** [comment]
I'm guessing that's fixed-by-design, we wouldn't be able to release a new version of the SDK without this change already being in :)

**Comment by kevmoo** [comment]
Yup. But I could still have a package with SDK `>=2.0.0 <3.0.0` that could be used by someone w/ an old SDK

**Comment by matanlurey** [comment]
Ok. I don't understand the scenario where the lint can apply on a user with an older SDK since the lint would only exist in the newer SDK. But Ok!

**Comment by kevmoo** [comment]
> Ok. I don't understand the scenario where the lint can apply on a user with an older SDK since the lint would only exist in the newer SDK. But Ok!

Me, working on `pkg:async` on Dart 2.1. Drop a bunch of imports to `dart:async`. Publish v2.0.1. Don't change SDK constraint.

Someone (using Flutter?) does pub upgrade, gets `pkg:async` v2.0.1 even though I'm on SDK 2.0.0-flutter.

Everything breaks.

**Comment by matanlurey** [comment]
Ok so that is a separate concern - we should probably let the rest of the team know to be careful when relying on `dart:core` for `Future` or `Stream`. @srawlins @lrhn send an email or two?

**Comment by jakemac53** [comment]
External package authors could very easily fall into this trap too, imo this seems like enough of an issue that we should probably block releasing this until we can get a validator implemented in pub that won't let you publish if you are using Future/Stream without the dart:async import and don't have the proper sdk constraint.

Even if you are following along with this feature, its super easy to accidentally do this.

**Comment by lrhn** [comment]
I don't think this is a new issue - any change to the SDK means that you must update your SDK requirement if you use it. This change isn't different from any other change we have made, it just happens to be the first non-bugfix change after the 2.0 release.

We should probably make it easy to test your package against an older SDK, say by having a `pub try` option that uploads the package and runs the tests with the oldest allowed SDK. It should also be able to tell you if your transitive requirements causes your effective SDK version to be higher than what you wrote.

**Comment by jakemac53** [comment]
> I don't think this is a new issue - any change to the SDK means that you must update your SDK requirement if you use it. This change isn't different from any other change we have made, it just happens to be the first non-bugfix change after the 2.0 release.

The difference is how easy this is to accidentally do, migrate a single method somewhere to use `async` and change its return type to a Future, and you have now broken all users of your package on old sdks. It is an incredibly easy trap to fall into.

**Comment by bwilkerson** [comment]
> We should probably make it easy to test your package against an older SDK ...

I agree that that would be a good tool to have, but the concern here isn't "your" package, it's the packages that depend on your package.

It sounds like we need a way for our tools to prevent code that depends on using one set of versions of the SDK from being run on other versions of the SDK (whether newer or older), where the set of versions is the intersection of the versions specified for each of the packages selected by `pub`. As far as I know we don't have that today.

**Comment by jakemac53** [comment]
> It sounds like we need a way for our tools to prevent code that depends on using one set of versions of the SDK from being run on other versions of the SDK (whether newer or older), where the set of versions is the intersection of the versions specified for each of the packages selected by pub. As far as I know we don't have that today.

We have this already (pub get/pub run do this).

The issue here is that this particular change makes it very easy to _accidentally_ introduce an sdk dependency _without realizing it_. It is up to package authors to update their minimum sdk manually, which is usually not a huge problem (although it does cause issues occasionally).

**Comment by srawlins** [comment]
The unnecessary_import Hint reports this today.

It is unaware of whether SDK constraints or other package constraints fit the diagnostic. The version constraints would not just be a problem for the SDK, but also for packages. Since an import directive can only be considered unnecessary because of an export directive somewhere, we could take advantage of a `@Since` annotation on the export.
